docs/spec/dataclasses.rst

```diff
@@ -0,0 +1,517 @@
+
+  # Using keyword arguments
+  c2 = CustomerModel(id=327, name="John Smith")
+
+  # These calls will generate runtime errors and should be flagged as
+  # errors by a static type checker.
+  c3 = CustomerModel()
+  c4 = CustomerModel(327, first_name="John")
+  c5 = CustomerModel(327, "John Smith", 0)
+
+Decorator function example
+""""""""""""""""""""""""""
+
+.. code-block:: python
+
+  _T = TypeVar("_T")
+  
+  # The ``create_model`` decorator is defined by a library.
+  # This could be in a type stub or inline.
+  @typing.dataclass_transform()
+  def create_model(cls: Type[_T]) -> Type[_T]:
+      cls.__init__ = ...
+      cls.__eq__ = ...
+      cls.__ne__ = ...
+      return cls
+  
+  # The ``create_model`` decorator can now be used to create new model
+  # classes, like this:
+  @create_model
+  class CustomerModel:
+      id: int
+      name: str
+
+Class example
+"""""""""""""
+
+.. code-block:: python
+
+  # The ``ModelBase`` class is defined by a library. This could be in
+  # a type stub or inline.
+  @typing.dataclass_transform()
+  class ModelBase: ...
+
+  # The ``ModelBase`` class can now be used to create new model
+  # subclasses, like this:
+  class CustomerModel(ModelBase):
+      id: int
+      name: str
+
+Metaclass example
+"""""""""""""""""
+
+.. code-block:: python
+
+  # The ``ModelMeta`` metaclass and ``ModelBase`` class are defined by
+  # a library. This could be in a type stub or inline.
+  @typing.dataclass_transform()
+  class ModelMeta(type): ...
+  
+  class ModelBase(metaclass=ModelMeta): ...
+  
+  # The ``ModelBase`` class can now be used to create new model
+  # subclasses, like this:
+  class CustomerModel(ModelBase):
+      id: int
+      name: str
+
+Decorator function and class/metaclass parameters
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A decorator function, class, or metaclass that provides dataclass-like
+functionality may accept parameters that modify certain behaviors.
+This specification defines the following parameters that static type
+checkers must honor if they are used by a dataclass transform. Each of
+these parameters accepts a bool argument, and it must be possible for
+the bool value (``True`` or ``False``) to be statically evaluated.
+
+* ``eq``,  ``order``, ``frozen``, ``init`` and ``unsafe_hash`` are parameters
+  supported in the stdlib dataclass, with meanings defined in 
+  :pep:`PEP 557 <557#id7>`.
+* ``kw_only``, ``match_args`` and ``slots`` are parameters supported
+  in the stdlib dataclass, first introduced in Python 3.10.
+
+``dataclass_transform`` parameters
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Parameters to ``dataclass_transform`` allow for some basic
+customization of default behaviors:
+
+.. code-block:: python
+
+  _T = TypeVar("_T")
+  
+  def dataclass_transform(
+      *,
+      eq_default: bool = True,
+      order_default: bool = False,
+      kw_only_default: bool = False,
+      field_specifiers: tuple[type | Callable[..., Any], ...] = (),
```

Comment:
New python versions also have `frozen_default`, should it be included?

Reply:
Good point, we should add that. (I'm also missing PEP 702, which I think got accepted after I started working on the spec).